### PR TITLE
Add User Auth and Athorization

### PR DESCRIPTION
**Pull request by JawadHossain**
Add user login, signup, auth, and route authorization.
Form Styling to be added in another pr so that this one does not get too big.

### CODE REVIEW

File: code/socialDistribution/views.py [hunk 1]
```diff
@@ -1,13 +1,82 @@
...
+
+    if request.method == 'POST':
+        form = CreateUserForm(request.POST)
+        if form.is_valid():
+            user = form.save()
+            # extract form data
+            username = form.cleaned_data.get('username')
+            first_name = form.cleaned_data.get('first_name')
+            last_name = form.cleaned_data.get('last_name')
+            github_url = request.POST.get('github_url', '')
```
Comment (tylerelston) [inline code/socialDistribution/views.py:49]: Using an incorrect github url, I got a 500 error but the registration still went through. Should we do some error handling?

File: code/socialDistribution/templates/navBase.html [hunk 2]
```diff
@@ -24,11 +24,28 @@
...
+
       <div class="collapse navbar-collapse" id="navbarSupportedContent">
-        <form class="mx-auto w-25">
-          <input class="form-control me-2" type="search" placeholder="Search" aria-label="Search">
-        </form>
-        <button class="btn btn-secondary" type="logout">Logout</button>
+        <!-- Left Nav Links -->
+        <ul class="navbar-nav me-auto mb-2 mb-lg-0">
+          <li class="nav-item">
+            <a class="nav-link active" aria-current="page" href="{% url 'logout' %}">Home</a>
```
Comment (tylerelston) [inline code/socialDistribution/templates/navBase.html:32]: We can remove the home link

File: code/socialDistribution/middleware.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    def __call__(self, request):
+        response = self.get_response(request)
+        return response
+    
+    def process_view(self, request, view_func, view_args, view_kwargs):
+        assert hasattr(request, 'user')
+        path = request.path_info.lstrip('/')
+
+        print(path)
```
[the remark this answers is not in the document]
Comment (JawadHossain) [reply code/socialDistribution/middleware.py]: Yes, thanks

File: code/socialDistribution/views.py (hunk 1 above)
Comment (JawadHossain) [reply code/socialDistribution/views.py]: Added GitHub URL check. Thanks.

File: code/socialDistribution/views.py [hunk 4]
```diff
@@ -1,13 +1,93 @@
...
+                # check github url
+                if (github_url and not github_url.startswith('https://github.com/')):
+                    context = { 'form': form }
+                    messages.info(request, 'Invalid github url, must be of format: https://github.com/username')
+                    return render(request, 'user/register.html', context)
+
+                user = form.save()
+
+                # add user to author group by default
+                group = Group.objects.get(name="author")
```
Comment (tylerelston) [inline code/socialDistribution/views.py:61]: Should this line be `group, created = Group.objects.get_or_create(name="author")`?
I get a 500 error if only `get` is used

**Review by debecerra — APPROVED** [review]

**Pull request merged**